File: python/day8.py
CHARS = ['a', 'b', 'c', 'd', 'e', 'f', 'g']

def parse_pattern(w):
    # represent e.g. the word acf as the tuple
    # (1,0,1,0,0,1,0)
    # (a,b,c,d,e,f,g)
    # this representation will be called binary pattern in what follows

    res = [0] * len(CHARS)
    for i, c in enumerate(CHARS):
        if c in w:
            res[i] = 1

    return tuple(res)

def permute(pi, bin_pattern):
    res = [0] * len(CHARS)

    for i, b in enumerate(bin_pattern):
        res[pi[i]] = b

    return tuple(res)

def solve():
    DIGITS = [
        'abcefg', #0
        'cf', #1
        'acdeg', #2
        'acdfg', #3
        'bcdf', #4
        'abdfg', #5
        'abdefg', #6
        'acf', #7
        'abcdefg', #8
        'abcdfg' #9
    ]

    with open("problems/problem8") as f:
        lines = f.readlines()

    observations = []
    for line in lines:
        i, o = line.split(' | ')

        i = [parse_pattern(w) for w in i.split()]
        o = [parse_pattern(w) for w in o.split()]

        observations.append((i, o))

    # This is not the most efficient solution for part 1 but it does not really matter
    prob1 = 0
    target_lengths = {len(DIGITS[1]),len(DIGITS[4]), len(DIGITS[7]), len(DIGITS[8])}
    for _, output in observations:
        for pattern in output:
            ones = sum(pattern)
            if  ones in target_lengths:
                prob1 += 1

    print("Problem 1:", prob1)

    ### Part 2

    # Now, our task for part 2 can be stated like this: Find a permutation pi of {0, 1,...,6} such that
    # for every binary input pattern (x0, ..., x6) the binary pattern (x)^pi = (x_pi^-1(0), x_pi^-1(1), ..., x_pi^-1(6))
    # is a valid binary pattern, i.e. a key in bin_pattern_to_digit.
    # Note that it is somewhat more natural to work with the inverse of the permutation in this instance:
    # If pi(i) = j then the jth component of the permuted tuple equals x_pi^-1{j} = x_i

    bin_pattern_to_digit = {parse_pattern(w):i for i, w in enumerate(DIGITS)}

    # We can also reduce the search space using a simple observation: 
    # Suppose that (x0, ..., x6) is an input pattern with exactly 3 one entries.
    # Then it has to represent a 7 and it holds that
    # (x)^pi = (x_pi^-1(0), ..., x_pipi^-1(6)) = (1, 0, 1, 0, 0, 1, 0)
    # If x_j = 1 then the pi(j) th component of the permuted vector is also 1, so that
    # pi(j) \in {0, 2, 5} holds.
    possible_positions = {i:set() for i in range(len(CHARS) + 1)}
    for bin_pattern in bin_pattern_to_digit:
        ones = sum(bin_pattern)
        one_indices = (i for i, x in enumerate(bin_pattern) if x == 1)

        possible_positions[ones].update(one_indices)
    
    prob2 = 0
    for inp, outp in observations:
        possible_pi = {i:set(range(len(CHARS))) for i in range(len(CHARS))}

        for bin_pattern in inp:
            l = sum(bin_pattern)

            for i, b in enumerate(bin_pattern):
                if not b:
                    continue

                possible_pi[i] &= possible_positions[l]

        # We find this permutation using backtracking
        def make_pi(pi, i, used):
            if i == len(CHARS):
                if len(used) == len(CHARS):
                    # pi is an actual permutation! Now verify its what we are looking for.
                    for w in inp:
                        if permute(pi, w) not in bin_pattern_to_digit:
                            return None

                    return pi
                else:
                    return None

            for j in possible_pi[i]:
                if j in used:
                    continue

                used.add(j)
                pi[i] = j
                
                res = make_pi(pi, i + 1, used)

                if res:
                    return res

                pi[i] = None
                used.remove(j)

        pi = make_pi({i:None for i in range(len(CHARS))}, 0, set())

        displayed_number = 0
        for w in outp:
            displayed_number *= 10
            displayed_number += bin_pattern_to_digit[permute(pi, w)]

        prob2 += displayed_number

    print("Problem 2:", prob2)

if __name__ == '__main__':
    solve()

File: src/day8.rs
use std::collections::{HashMap, HashSet};

const INPUT: &str = include_str!("../problems/problem8");
const N: usize = 7;
const CHARS: [char; N] = ['a', 'b', 'c', 'd', 'e', 'f', 'g'];
const DIGIT_PATTERNS: [&str; 10] = [
    "abcefg", "cf", "acdeg", "acdfg", "bcdf", "abdfg", "abdefg", "acf", "abcdefg", "abcdfg",
];

type Permutation = [u8; N];
type Pattern = [u8; N];

pub fn parse_pattern(s: &str) -> Pattern {
    let mut ret = [0; N];

    for i in 0..N {
        if s.contains(CHARS[i]) {
            ret[i] = 1;
        }
    }

    ret
}

pub fn permute(pattern: &Pattern, pi: &Permutation) -> Pattern {
    let mut ret = [0; N];

    for i in 0..N {
        ret[pi[i] as usize] = pattern[i]
    }

    ret
}

pub fn solve() -> crate::Result<()> {
    let observations = INPUT
        .lines()
        .map(|line| {
            let mut split = line.split(" | ").map(|part| {
                part.split_whitespace()
                    .map(|pat| parse_pattern(pat))
                    .collect::<Vec<_>>()
            });

            (split.next().unwrap(), split.next().unwrap())
        })
        .collect::<Vec<_>>();

    // Determine a permutation pi such that for every input binary pattern
    // (x6, ..., x0) it holds that (x_pi^-1(6), ..., x^pi^-1(0)) is a valid pattern
    fn make_pi<F: Fn(&Pattern) -> bool>(
        mut pi: Permutation,
        i: usize,
        used: &mut HashSet<u8>,
        inp: &Vec<Pattern>,
        valid_pattern: &F,
    ) -> Option<Permutation> {
        if i == N {
            if used.len() < N {
                return None;
            }

            for pattern in inp {
                let permuted = permute(pattern, &pi);

                if !valid_pattern(&permuted) {
                    return None;
                }
            }

            Some(pi)
        } else {
            for j in 0..(N as u8) {
                if used.contains(&j) {
                    continue;
                }

                used.insert(j);
                pi[i] = j;

                if let Some(pi) = make_pi(pi, i + 1, used, inp, valid_pattern) {
                    return Some(pi);
                }

                used.remove(&j);
            }

            None
        }
    }

    let digit_patterns: HashMap<Pattern, u8> = DIGIT_PATTERNS
        .iter()
        .enumerate()
        .map(|(i, s)| (parse_pattern(s), i as u8))
        .collect();

    let mut prob1 = 0;
    let mut prob2 = 0;

    for (inp, outp) in observations {
        let pi = make_pi([0; N], 0, &mut Default::default(), &inp, &|p: &Pattern| {
            digit_patterns.contains_key(p)
        })
        .ok_or(crate::Error::InvalidInput)?;

        let mut tmp: u64 = 0;
        for w in outp {
            tmp *= 10;
            tmp += digit_patterns[&permute(&w, &pi)] as u64;

            match w.iter().sum() {
                2 | 3 | 4 | 7 => prob1 += 1,
                _ => {}
            }
        }

        prob2 += tmp;
    }

    println!("Problem 1: {}", prob1);
    println!("Problem 2: {}", prob2);

    Ok(())
}
